File: R/fn-model.R
#### run with priorTempered phase 1 --------------------------------------------
library(gpds)
# set up configuration if not already exist ------------------------------------
if(!exists("config")){
  config <- list(
    nobs = 26,
    noise = c(0.1, 0.1),
    kernel = "generalMatern",
    seed = 1365546660, #(as.integer(Sys.time())*104729+sample(1e9,1))%%1e9,
    npostplot = 50,
    loglikflag = "withmeanBand",
    bandsize = 20,
    hmcSteps = 500,
    n.iter = 1e4,
    burninRatio = 0.50,
    stepSizeFactor = 1,
    filllevel = 2,
    modelName = "FN",
    startXAtTruth = FALSE,
    startThetaAtTruth = FALSE,
    startSigmaAtTruth = FALSE,
    useGPmean = TRUE,
    forseTrueMean = FALSE,
    phase2 = FALSE,
    temperPrior = TRUE,
    phase3 = FALSE,
    max.epoch = 10,
    epoch_method = c("mean", "median", "deSolve")[3]
  )
}

config$ndis <- (config$nobs-1)*2^config$filllevel+1
if(config$temperPrior){
  config$priorTemperature <- config$ndis / config$nobs  
}else{
  config$priorTemperature <- 1
}

# initialize global parameters, true x, simulated x ----------------------------
if(grepl("/n/",getwd())){
  baseDir <- "/n/regal/kou_lab/shihaoyang/DynamicSys/results/" # tmp folder on cluster 
  config$seed <- (as.integer(Sys.time())*104729+sample(1e9,1))%%1e9 # random seed on cluster
}else{
  baseDir <- "~/Workspace/DynamicSys/results/batch-output/"  
}
outDir <- with(config, paste0(baseDir, modelName, "-", loglikflag,"-", kernel,
                              "-nobs",nobs,"-noise", paste(round(noise,3), collapse = "_"),
                              "-ndis",ndis,"-",ifelse(config$temperPrior, "temperPrior", "unitHeatPrior"),"/"))
system(paste("mkdir -p", outDir))

pram.true <- list(
  theta=c(0.2,0.2,3),
  x0 = c(-1, 1),
  phi=c(0.9486433, 3.2682434,
         1.9840824, 1.1185157),
  sigma=config$noise
)

times <- seq(0,20,0.05)

modelODE <- function(t, state, parameters) {
  list(as.vector(gpds:::fnmodelODE(parameters, t(state))))
}

xtrue <- deSolve::ode(y = pram.true$x0, times = times, func = modelODE, parms = pram.true$theta)
xtrue <- data.frame(xtrue)
matplot(xtrue[, "time"], xtrue[, -1], type="l", lty=1)

xtrueFunc <- lapply(2:ncol(xtrue), function(j)
  approxfun(xtrue[, "time"], xtrue[, j]))

xsim <- xtrue

set.seed(config$seed)
for(j in 1:(ncol(xsim)-1)){
  xsim[,1+j] <- xsim[,1+j]+rnorm(nrow(xsim), sd=config$noise[j])  
}

xsim.obs <- xsim[seq(1,nrow(xsim), length=config$nobs),]
matplot(xsim.obs$time, xsim.obs[,-1], type="p", col=1:(ncol(xsim)-1), pch=20, add = TRUE)

matplot(xsim.obs$time, xsim.obs[,-1], type="p", col=1:(ncol(xsim)-1), pch=20)

xsim <- insertNaN(xsim.obs,config$filllevel)

tvec.full <- xsim$time
tvec.nobs <- xsim.obs$time

foo <- outer(tvec.full, t(tvec.full),'-')[,1,]
r <- abs(foo)
r2 <- r^2
signr <- -sign(foo)

foo <- outer(tvec.nobs, t(tvec.nobs),'-')[,1,]
r.nobs <- abs(foo)
r2.nobs <- r.nobs^2
signr.nobs <- -sign(foo)

# GPsmoothing: marllik+fftNormalprior for phi-sigma ----------------------------
cursigma <- rep(NA, ncol(xsim)-1)
curphi <- matrix(NA, 2, ncol(xsim)-1)

for(j in 1:(ncol(xsim)-1)){
  priorFactor <- getFrequencyBasedPrior(xsim.obs[,1+j])
  
  desiredMode <- priorFactor["meanFactor"]
  betaRate <- uniroot(function(betaRate) pgamma(1, 1 + desiredMode*betaRate, betaRate)-0.95,
                      c(1e-3, 1e3))$root
  alphaRate <- 1 + desiredMode*betaRate
  
  fn <- function(par) {
    marlik <- phisigllikC( par, data.matrix(xsim.obs[,1+j]), r.nobs, config$kernel)
    penalty <- dnorm(par[2], max(xsim.obs$time)*priorFactor["meanFactor"], 
                     max(xsim.obs$time)*priorFactor["sdFactor"], log=TRUE)
    # penalty <- dgamma(par[2], alphaRate, betaRate/max(xsim.obs$time), log=TRUE)
    # penalty <- 0
    -(marlik$value + penalty)
  }
  gr <- function(par) {
    marlik <- phisigllikC( par, data.matrix(xsim.obs[,1+j]), r.nobs, config$kernel)
    grad <- -as.vector(marlik$grad)
    penalty <- (par[2] - max(xsim.obs$time)*priorFactor["meanFactor"]) / (max(xsim.obs$time)*priorFactor["sdFactor"])^2
    # penalty <- ((alphaRate-1)/par[2] - betaRate/max(xsim.obs$time))
    # penalty <- 0
    grad[2] <- grad[2] + penalty
    grad
  }
  testthat::expect_equal(gr(c(5,50,1))[2], (fn(c(5,50+1e-6,1)) - fn(c(5,50,1)))/1e-6, tolerance=1e-3)
  marlikmap <- optim(c(sd(xsim.obs[,1+j])/2, max(xsim.obs$time)/2, sd(xsim.obs[,1+j])/2), 
                     fn, gr, method="L-BFGS-B", lower = 0.0001,
                     upper = c(Inf, Inf, Inf))
  
  cursigma[j] <- marlikmap$par[3]
  curphi[,j] <- marlikmap$par[1:2]
}

cursigma
curphi

curCov <- lapply(1:(ncol(xsim.obs)-1), function(j){
  covEach <- calCov(curphi[, j], r, signr, bandsize=config$bandsize, 
                    kerneltype=config$kernel)
  covEach$mu[] <- mean(xsim.obs[,j+1])
  covEach
})

if(config$useGPmean){
  for(j in 1:(ncol(xsim)-1)){
    ydy <- getMeanCurve(xsim.obs$time, xsim.obs[,j+1], xsim$time, 
                        t(curphi[,j]), t(cursigma[j]), 
                        kerneltype=config$kernel, deriv = TRUE)
    curCov[[j]]$mu <- as.vector(ydy[[1]])
    curCov[[j]]$dotmu <- as.vector(ydy[[2]])
  }
}

gpsmoothFuncList <- list()
for(j in 1:(ncol(xsim)-1)){
  ynew <- getMeanCurve(xsim.obs$time, xsim.obs[,j+1], xsim$time, 
                       t(curphi[,j]), t(cursigma[j]), kerneltype=config$kernel)
  gpsmoothFuncList[[j]] <- approxfun(xsim$time, ynew)
  plot.function(gpsmoothFuncList[[j]], from = min(xsim$time), to = max(xsim$time),
                lty = 2, col = j, add = TRUE)
}
cursigma
# MCMC starting value ----------------------------------------------------------
yobs <- data.matrix(xsim[,-1])

xsimInit <- xsim
for(j in 1:(ncol(xsim)-1)){
  nanId <- which(is.na(xsimInit[,j+1]))
  xsimInit[nanId,j+1] <- gpsmoothFuncList[[j]](xsimInit$time[nanId])
}
matplot(xsimInit$time, xsimInit[,-1], type="p", pch=2, add=TRUE)
xInit <- data.matrix(xsimInit[,-1])

thetaInit <- rep(1, length(pram.true$theta))

thetaoptim <- function(xInit, thetaInit, curphi, cursigma){
  curCov <- lapply(1:(ncol(xsim.obs)-1), function(j){
    covEach <- calCov(curphi[, j], r, signr, bandsize=config$bandsize, 
                      kerneltype=config$kernel)
    covEach$mu[] <- mean(xsim.obs[,j+1])
    covEach
  })
  fn <- function(par) {
    -xthetallikRcpp( yobs, curCov, cursigma, c(xInit, par), "FN" )$value
  }
  gr <- function(par) {
    -as.vector(xthetallikRcpp( yobs, curCov, cursigma, c(xInit, par), "FN" )$grad[-(1:length(xInit))])
  }
  marlikmap <- optim(c(thetaInit), fn, gr, 
                     method="L-BFGS-B", lower = 0.001, control = list(maxit=1e5))
  thetaInit[] <- marlikmap$par
  list(thetaInit = thetaInit)
}

thetamle <- thetaoptim(xInit, thetaInit, curphi, cursigma)
thetaInit <- thetamle$thetaInit
sigmaInit <- pmax(cursigma, 0.01)

if(config$startXAtTruth){
  xInit <- sapply(xtrueFunc, function(f) f(xsim$time))  
}
if(config$startThetaAtTruth){
  thetaInit <- pram.true$theta  
}
if(config$startSigmaAtTruth){
  sigmaInit <- pram.true$sigma  
}

xthetasigmaInit <- c(xInit, thetaInit, sigmaInit)
stepLowInit <- rep(0.000035, length(xthetasigmaInit))
stepLowInit <- stepLowInit*config$stepSizeFactor

if(config$forseTrueMean){
  dotxtrue = gpds:::fnmodelODE(pram.true$theta, data.matrix(xtrue[,-1]))
  dotxtrue.atsim <- dotxtrue[round(xtrue[,"time"],3) %in% round(xsim$time,3),]
  xtrue.atsim <- xtrue[round(xtrue[,"time"],3) %in% round(xsim$time,3),-1]
  for(j in 1:2){
    curCov[[j]]$mu <- xtrue.atsim[, j]
    curCov[[j]]$dotmu <- dotxtrue.atsim[, j]
  }
}

# HMC sampler for x, theta, sigma ----------------------------------------------
xId <- 1:length(xInit)
thetaId <- (max(xId)+1):(max(xId)+length(thetaInit))
sigmaId <- (max(thetaId)+1):(max(thetaId)+length(sigmaInit))

xthetasigamSingleSampler <- function(xthetasigma, stepSize) 
  xthetasigmaSample(yobs, curCov, xthetasigma[sigmaId], xthetasigma[c(xId, thetaId)], 
                    stepSize, config$hmcSteps, F, loglikflag = config$loglikflag,
                    priorTemperature = config$priorTemperature, modelName = config$modelName)
# stepLowInit[sigmaId] <- 0
chainSamplesOut <- chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit, verbose=TRUE)

burnin <- as.integer(config$n.iter*config$burninRatio)
gpode <- list(theta=chainSamplesOut$xth[-(1:burnin), thetaId],
              xsampled=array(chainSamplesOut$xth[-(1:burnin), xId], 
                             dim=c(config$n.iter-burnin, nrow(xsim), ncol(xsim)-1)),
              lglik=chainSamplesOut$lliklist[-(1:burnin)],
              sigma = chainSamplesOut$xth[-(1:burnin), sigmaId, drop=FALSE],
              phi = curphi)
gpode$fode <- sapply(1:length(gpode$lglik), function(t) 
  with(gpode, gpds:::fnmodelODE(theta[t,], xsampled[t,,])), simplify = "array")
gpode$fode <- aperm(gpode$fode, c(3,1,2))

dotxtrue = gpds:::fnmodelODE(pram.true$theta, data.matrix(xtrue[,-1]))

configWithPhiSig <- config
philist <- lapply(data.frame(round(curphi,3)), function(x) paste(x, collapse = "; "))
names(philist) <- paste0("phi", 1:length(philist))
configWithPhiSig <- c(configWithPhiSig, philist)

odemodel <- list(times=times, modelODE=modelODE, xtrue=xtrue, curCov=curCov)

gpds:::plotPostSamplesFlex(
  paste0(outDir, config$kernel,"-",config$seed,"-priorTempered-phase1.pdf"), 
  xtrue, dotxtrue, xsim, gpode, pram.true, configWithPhiSig, odemodel)

absCI <- apply(gpode$theta, 2, quantile, probs = c(0.025, 0.5, 0.975))
absCI <- rbind(absCI, mean=colMeans(gpode$theta))
absCI <- rbind(absCI, coverage = (absCI["2.5%",] < pram.true$theta &  pram.true$theta < absCI["97.5%",]))

saveRDS(absCI, paste0(
  outDir,
  config$loglikflag,"-priorTempered-phase1-",config$kernel,"-",config$seed,".rds"))

save.image(paste0(
  outDir,
  config$loglikflag,"-priorTempered-phase",1,"-",config$kernel,"-",config$seed,".rda"))


# phase 2 ----------------------------------------------------------------------
if(config$phase2){
  muAllDim <- apply(gpode$xsampled, 2:3, mean)
  dotmuAllDim <- apply(gpode$fode, 2:3, mean) 
  
  for(j in 1:ncol(muAllDim)){
    curCov[[j]]$mu <- muAllDim[,j]
    curCov[[j]]$dotmu <- dotmuAllDim[,j]
  }
  
  xInit <- muAllDim
  thetaInit <- colMeans(gpode$theta)
  stepLowInit <- chainSamplesOut$stepLow
  config$burninRatio <- 0.5
  config$phase2 <- TRUE
  
  xthetasigamSingleSampler <- function(xthetasigma, stepSize) 
    xthetasigmaSample(yobs, curCov, xthetasigma[sigmaId], xthetasigma[c(xId, thetaId)], 
                      stepSize, config$hmcSteps, F, loglikflag = config$loglikflag,
                      priorTemperature = config$priorTemperature, modelName = config$modelName)
  # stepLowInit[sigmaId] <- 0
  chainSamplesOut <- chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit, verbose=TRUE)
  
  burnin <- as.integer(config$n.iter*config$burninRatio)
  gpode <- list(theta=chainSamplesOut$xth[-(1:burnin), thetaId],
                xsampled=array(chainSamplesOut$xth[-(1:burnin), xId], 
                               dim=c(config$n.iter-burnin, nrow(xsim), ncol(xsim)-1)),
                lglik=chainSamplesOut$lliklist[-(1:burnin)],
                sigma = chainSamplesOut$xth[-(1:burnin), sigmaId, drop=FALSE],
                phi = curphi)
  gpode$fode <- sapply(1:length(gpode$lglik), function(t) 
    with(gpode, gpds:::fnmodelODE(theta[t,], xsampled[t,,])), simplify = "array")
  gpode$fode <- aperm(gpode$fode, c(3,1,2))
  
  dotxtrue = gpds:::fnmodelODE(pram.true$theta, data.matrix(xtrue[,-1]))
  
  configWithPhiSig <- config
  philist <- lapply(data.frame(round(curphi,3)), function(x) paste(x, collapse = "; "))
  names(philist) <- paste0("phi", 1:length(philist))
  configWithPhiSig <- c(configWithPhiSig, philist)
  
  odemodel <- list(times=times, modelODE=modelODE, xtrue=xtrue, curCov=curCov)
  
  gpds:::plotPostSamplesFlex(
    paste0(outDir, config$kernel,"-",config$seed,"-priorTempered-phase2.pdf"), 
    xtrue, dotxtrue, xsim, gpode, pram.true, configWithPhiSig, odemodel)
  
  absCI <- apply(gpode$theta, 2, quantile, probs = c(0.025, 0.5, 0.975))
  absCI <- rbind(absCI, mean=colMeans(gpode$theta))
  absCI <- rbind(absCI, coverage = (absCI["2.5%",] < pram.true$theta &  pram.true$theta < absCI["97.5%",]))
  
  saveRDS(absCI, paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase2-",config$kernel,"-",config$seed,".rds"))
  
  save.image(paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase",2,"-",config$kernel,"-",config$seed,".rda"))
  
}

# phase 3 ----------------------------------------------------------------------
if(config$phase3){
  muAllDim <- apply(gpode$xsampled, 2:3, mean)
  dotmuAllDim <- apply(gpode$fode, 2:3, mean) 
  
  for(j in 1:ncol(muAllDim)){
    curCov[[j]]$mu <- muAllDim[,j]
    curCov[[j]]$dotmu <- dotmuAllDim[,j]
  }
  
  xInit <- muAllDim
  thetaInit <- colMeans(gpode$theta)
  stepLowInit <- chainSamplesOut$stepLow
  config$burninRatio <- 0.5
  config$phase3 <- TRUE
  
  xthetasigamSingleSampler <- function(xthetasigma, stepSize) 
    xthetasigmaSample(yobs, curCov, xthetasigma[sigmaId], xthetasigma[c(xId, thetaId)], 
                      stepSize, config$hmcSteps, F, loglikflag = config$loglikflag,
                      priorTemperature = config$priorTemperature, modelName = config$modelName)
  # stepLowInit[sigmaId] <- 0
  chainSamplesOut <- chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit, verbose=TRUE)
  
  burnin <- as.integer(config$n.iter*config$burninRatio)
  gpode <- list(theta=chainSamplesOut$xth[-(1:burnin), thetaId],
                xsampled=array(chainSamplesOut$xth[-(1:burnin), xId], 
                               dim=c(config$n.iter-burnin, nrow(xsim), ncol(xsim)-1)),
                lglik=chainSamplesOut$lliklist[-(1:burnin)],
                sigma = chainSamplesOut$xth[-(1:burnin), sigmaId, drop=FALSE],
                phi = curphi)
  gpode$fode <- sapply(1:length(gpode$lglik), function(t) 
    with(gpode, gpds:::fnmodelODE(theta[t,], xsampled[t,,])), simplify = "array")
  gpode$fode <- aperm(gpode$fode, c(3,1,2))
  
  dotxtrue = gpds:::fnmodelODE(pram.true$theta, data.matrix(xtrue[,-1]))
  
  configWithPhiSig <- config
  philist <- lapply(data.frame(round(curphi,3)), function(x) paste(x, collapse = "; "))
  names(philist) <- paste0("phi", 1:length(philist))
  configWithPhiSig <- c(configWithPhiSig, philist)
  
  odemodel <- list(times=times, modelODE=modelODE, xtrue=xtrue, curCov=curCov)
  
  gpds:::plotPostSamplesFlex(
    paste0(outDir, config$kernel,"-",config$seed,"-priorTempered-phase3.pdf"), 
    xtrue, dotxtrue, xsim, gpode, pram.true, configWithPhiSig, odemodel)
  
  absCI <- apply(gpode$theta, 2, quantile, probs = c(0.025, 0.5, 0.975))
  absCI <- rbind(absCI, mean=colMeans(gpode$theta))
  absCI <- rbind(absCI, coverage = (absCI["2.5%",] < pram.true$theta &  pram.true$theta < absCI["97.5%",]))
  
  saveRDS(absCI, paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase3-",config$kernel,"-",config$seed,".rds"))
  
  save.image(paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase",3,"-",config$kernel,"-",config$seed,".rda"))
  
}

# epoch ----------------------------------------------------------------------------------
config$burninRatio <- 0.2
config$n.iter <- 5000

for(epoch in 4:config$max.epoch){
  if(config$epoch_method == "mean"){
    muAllDim <- apply(gpode$xsampled, 2:3, mean)
    dotmuAllDim <- apply(gpode$fode, 2:3, mean) 
  }else if(config$epoch_method == "median"){
    muAllDim <- apply(gpode$xsampled, 2:3, median)
    dotmuAllDim <- apply(gpode$fode, 2:3, median) 
  }else if(config$epoch_method == "deSolve"){
    thetaInit <- apply(gpode$theta, 2, mean)
    x0Init <- apply(gpode$xsampled, 2:3, mean)[1,]
    
    ttheta <- colMeans(gpode$theta)
    tx0 <- colMeans(gpode$xsampled[,1,])
    xdesolvePM <- deSolve::ode(y = tx0, times = xsim$time, func = modelODE, parms = ttheta)
    muAllDim <- as.matrix(xdesolvePM[,-1])
    dotmuAllDim <- gpds:::fnmodelODE(ttheta, muAllDim)
  }
  
  for(j in 1:ncol(muAllDim)){
    curCov[[j]]$mu <- muAllDim[,j]
    curCov[[j]]$dotmu <- dotmuAllDim[,j]
  }
  
  xInit <- muAllDim
  thetaInit <- apply(gpode$theta, 2, mean)
  stepLowInit <- chainSamplesOut$stepLow
  
  xthetasigamSingleSampler <- function(xthetasigma, stepSize) 
    xthetasigmaSample(yobs, curCov, xthetasigma[sigmaId], xthetasigma[c(xId, thetaId)], 
                      stepSize, config$hmcSteps, F, loglikflag = config$loglikflag,
                      priorTemperature = config$priorTemperature, modelName = config$modelName)
  # stepLowInit[sigmaId] <- 0
  chainSamplesOut <- chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit, verbose=TRUE)
  
  burnin <- as.integer(config$n.iter*config$burninRatio)
  gpode <- list(theta=chainSamplesOut$xth[-(1:burnin), thetaId],
                xsampled=array(chainSamplesOut$xth[-(1:burnin), xId], 
                               dim=c(config$n.iter-burnin, nrow(xsim), ncol(xsim)-1)),
                lglik=chainSamplesOut$lliklist[-(1:burnin)],
                sigma = chainSamplesOut$xth[-(1:burnin), sigmaId, drop=FALSE],
                phi = curphi)
  gpode$fode <- sapply(1:length(gpode$lglik), function(t) 
    with(gpode, gpds:::fnmodelODE(theta[t,], xsampled[t,,])), simplify = "array")
  gpode$fode <- aperm(gpode$fode, c(3,1,2))
  
  dotxtrue = gpds:::fnmodelODE(pram.true$theta, data.matrix(xtrue[,-1]))
  
  configWithPhiSig <- config
  philist <- lapply(data.frame(round(curphi,3)), function(x) paste(x, collapse = "; "))
  names(philist) <- paste0("phi", 1:length(philist))
  configWithPhiSig <- c(configWithPhiSig, philist)
  
  odemodel <- list(times=times, modelODE=modelODE, xtrue=xtrue, curCov=curCov)
  
  gpds:::plotPostSamplesFlex(
    paste0(outDir, config$kernel,"-",config$seed,"-priorTempered-phase",epoch,".pdf"), 
    xtrue, dotxtrue, xsim, gpode, pram.true, configWithPhiSig, odemodel)
  
  absCI <- apply(gpode$theta, 2, quantile, probs = c(0.025, 0.5, 0.975))
  absCI <- rbind(absCI, mean=colMeans(gpode$theta))
  absCI <- rbind(absCI, coverage = (absCI["2.5%",] < pram.true$theta &  pram.true$theta < absCI["97.5%",]))
  
  saveRDS(absCI, paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase",epoch,"-",config$kernel,"-",config$seed,".rds"))
  save.image(paste0(
    outDir,
    config$loglikflag,"-priorTempered-phase",epoch,"-",config$kernel,"-",config$seed,".rda"))
}

File: matlabgpds/fn_model.m
config.nobs = 41;
config.noise = [0.15, 0.07] * 2;
config.kernel = "generalMatern";
config.seed = 1365546660; %(as.integer(Sys.time())*104729+sample(1e9,1))%%1e9,
config.npostplot = 50;
config.loglikflag = "withmeanBand";
config.bandsize = 20;
config.hmcSteps = 500;
config.n_iter = 1e4;
config.burninRatio = 0.50;
config.stepSizeFactor = 1;
config.filllevel = 2;
config.modelName = "FN";
config.startXAtTruth = false;
config.startThetaAtTruth = false;
config.startSigmaAtTruth = false;
config.useGPmean = true;
config.forseTrueMean = false;
config.phase2 = false;
config.temperPrior = true;
config.phase3 = false;
config.max_epoch = 10;
config.epoch_method = ["mean", "median", "deSolve", "f_x_bar"];
config.epoch_method = config.epoch_method(1);


config.ndis = (config.nobs-1)*2^config.filllevel+1;
if (config.temperPrior)
  config.priorTemperature = config.ndis / config.nobs;
  config.priorString = 'temperPrior';
else
  config.priorTemperature = 1;
  config.priorString = 'unitHeatPrior';
end

% initialize global parameters, true x, simulated x ----------------------------
baseDir = '/home/s246wong/dynsys-Matlab/';
outDir =  baseDir + config.modelName + '-' + config.loglikflag +  '-' + config.kernel ...
                              + '-nobs' + config.nobs +  '-noise' + num2str(config.noise, '%#5.3f_')...
                              + '-ndis' + config.ndis + '-' + config.priorString + '/';

system("mkdir -p " + outDir);

pram_true.theta = [0.2,0.2,3];
pram_true.x0 = [-1, 1];
pram_true.phi= [0.9486433, 3.2682434;
                1.9840824, 1.1185157];
pram_true.sigma=config.noise;

times = 0:(20/240):20;

%times = 0:0.00001:20;

[foo, xtrue]=ode45(@fnmodelODE,times,pram_true.x0,[],pram_true.theta);

xtrue = horzcat(foo,xtrue); 
plot(xtrue(:,1),xtrue(:,2:3));

xsim  = linspace(0,20,config.nobs);
xsim = horzcat(xsim',interp1(times,xtrue(:,2),xsim)',interp1(times,xtrue(:,3),xsim)');

rng(config.seed);

for j=1:(size(xsim,2)-1)
  xsim(:,1+j) = normrnd(xsim(:,1+j), config.noise(j));
end

xsim_obs = xsim( linspace(1, size(xsim,1), config.nobs),:);

hold on;
plot(xsim_obs(:,1), xsim_obs(:, 2:size(xsim,2)), 'o');
hold off;

plot(xsim_obs(:,1), xsim_obs(:, 2:size(xsim,2)), 'o');

xsim = insertNaN(xsim_obs,config.filllevel);

tvec.full = xsim(:,1);
tvec.nobs = xsim_obs(:,1);

foo1 = @(x,y)x-y;
foo = bsxfun(foo1, tvec.full, tvec.full');
r = abs(foo);
r2 = r.^2;
signr = -sign(foo);

foo = bsxfun(foo1, tvec.nobs, tvec.nobs');
r_nobs = abs(foo);
r2_nobs = r_nobs.^2;
signr_nobs = -sign(foo);

% GPsmoothing: marllik+fftNormalprior for phi-sigma ----------------------------
cursigma = NaN(1,size(xsim,2)-1);
curphi = NaN(2, size(xsim,2)-1);

for j=1:(size(xsim,2)-1)
    [priormeanFactor, priorsdFactor] = getFrequencyBasedPrior(xsim_obs(:,j+1));

    desiredMode = priormeanFactor;
    betaRate = fzero(@(betaRate) gamcdf(1, 1 + desiredMode*betaRate, 1/betaRate)-0.95, [1e-3 1e3]);
    alphaRate = 1 + desiredMode*betaRate;

    mlp = @(pars)marlikprior(pars, xsim_obs, r_nobs, config, j, priormeanFactor, priorsdFactor);
    optim_options = optimoptions(@fminunc,'Algorithm','trust-region','SpecifyObjectiveGradient',true);
    ostart = [std(xsim_obs(:,1+j))/2 max(xsim_obs(:,1))/2 std(xsim_obs(:,1+j))/2]; 
    [xout,fval,exitflag,output,grad] = fminunc(mlp,ostart,optim_options);
    
    cursigma(j) = xout(3);
    curphi(j,:) = xout(1:2);
end
%[t1 t2] = mlp([1 1]);

%j=1;
%covEach = calCov(curphi(:,j), r, signr, config.bandsize)
%calCov(curphi(:,j), r_nobs, signr_nobs, config.bandsize)

for j=1:(size(xsim,2)-1)
   %curCov(j) =  calCov(curphi(:,j), r_nobs, signr_nobs, config.bandsize);
   curCov(j) =  calCov(curphi(:,j), r, signr, config.bandsize);
end

if(config.useGPmean)
  for j=1:(size(xsim,2)-1)
    ydy = getMeanCurve(xsim_obs(:,1), xsim_obs(:,j+1), xsim(:,1), ...
                        curphi(:,j)', cursigma(j));
                        %kerneltype=config$kernel, deriv = TRUE);
    curCov(j).mu = ydy.y;
    curCov(j).dotmu = ydy.dy;
  end
end

% MCMC starting value ----------------------------------------------------------
yobs = xsim(:,2:size(xsim,2));

xsimInit = xsim;
for j=1:(size(xsim,2)-1)
  %nanId <- which(is.na(xsimInit[,j+1]))
  %xsimInit[nanId,j+1] <- gpsmoothFuncList[[j]](xsimInit$time[nanId])
  %xsimInit(:,j+1) = interp1( xsim_obs(:,1), xsim_obs(:,j+1), xsim(:,1), 'spline');
  xsimInit(:,j+1) = interp1( xsim_obs(:,1), xsim_obs(:,j+1), xsim(:,1), 'pchip');
end
%matplot(xsimInit$time, xsimInit[,-1], type="p", pch=2, add=TRUE)
xInit = reshape(xsimInit(:,2:size(xsim,2)),[1 size(xsim,1)*(size(xsim,2)-1)]);

thetaInit = ones(1, length(pram_true.theta));


%[ f1 g1 ] =  xthetallikM( yobs, curCov, cursigma, [xInit thetaInit], config.modelName, false );
chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit);
%[ f1 g1 ] = thetaoptim( yobs, curCov, cursigma, [xInit thetaInit], config.modelName, false );

llp = @(theta) thetaoptim(yobs, curCov, cursigma, [xInit theta], config.modelName, false );

%    optim_options = optimoptions(@fminunc,'Algorithm','trust-region','SpecifyObjectiveGradient',true);
    [xout,fval,exitflag,output,grad] = fminunc(llp,thetaInit,optim_options);
    thetaInit = xout;
    sigmaInit = cursigma;
xthetasigmaInit = [ xInit, thetaInit, sigmaInit ] ;
stepLowInit = ones(1, length(xthetasigmaInit)) * 0.000035;
stepLowInit = stepLowInit*config.stepSizeFactor;

xId = 1:length(xInit);
thetaId = (max(xId)+1):(max(xId)+length(thetaInit));
sigmaId = (max(thetaId)+1):(max(thetaId)+length(sigmaInit));


xthetasigamSingleSampler = @(xthetasigma, stepSize)  xthetasigmaSample(yobs, curCov, xthetasigma(sigmaId), xthetasigma([xId, thetaId]), ...
                    stepSize, config.hmcSteps, false, config.loglikflag, ...
                    config.priorTemperature, config.modelName);

chainSamplesOut = chainSampler(config, xthetasigmaInit, xthetasigamSingleSampler, stepLowInit);
    

